### ISSUE TITLE
Petastorm with LSTM model (slow performance)

### ISSUE THREAD

**Issue by jerrygb** [uber/petastorm#493]
Hi,

We are trying to Petastorm to stream dataset into one of models. We are using it to allow streaming dataset, since pandas was not option due to memory restrictions.

When using Petastorm even with smaller subset of the data, we seem to have poor performance. This is more than likely due to my configuration of the block size, parameters to the model or tensor we are trying to generate from the inferred schema object.

I would really appreciate your help in understanding the issue. Following is the code we have.

Data set is generated as follows,

```
df_splits = df.randomSplit([.7, .3], None)
df_splits[0].write.mode("overwrite").format("parquet").option("parquet.enable.summary-metadata", True).option("parquet.enable.summary-metadata", True).option("parquet.block.size", 512 * 1024 * 1024).option("compression", "gzip").mode("overwrite").save("tf_train.parquet.gz')
```
This seems to generate the same number of row groups irrespective of the block size setting. I am not sure if it relates to the bug [here](https://issues.apache.org/jira/browse/PARQUET-1337).

Data set is consumed by Petastorm as follows,

```
def parser(record, look_back, train_columns, batch_size):
    # logging.info('Length of Record: %d', len(record))
    features_cols = []
    labels_cols = []
    for col in train_columns:
        if col.find('LABEL_CLASSES') != 0: 
            features_cols.append(getattr(record, col)) # 9770 features
        else:
            labels_cols.append(getattr(record, col)) # 6 classes
    features = tf.concat(features_cols, axis=0)
    labels = tf.concat(labels_cols, axis=0)
    return tf.reshape(features, [batch_size, look_back, len(features_cols) // look_back]), \
           tf.reshape(labels, [batch_size, look_back, len(labels_cols) // look_back])

with make_batch_reader('file://{}'.format(os.path.join(training_input_dir, training_dataset_file)),
                       num_epochs=None,
                       cur_shard=hvd.rank(), shard_count=hvd.size()) as train_reader:
    with make_batch_reader(
            'file://{}'.format(os.path.join(training_input_dir, validation_dataset_file)),
            num_epochs=None,
            cur_shard=hvd.rank(), shard_count=hvd.size()) as validation_reader:

        train_ds = make_petastorm_dataset(train_reader) \
            .apply(tf.data.experimental.unbatch()) \
            .batch(batch_size_train) \
            .map(
            lambda x: parser(x, train_look_back, agg_relevant_columns_for_train, batch_size_train))
    validation_ds = make_petastorm_dataset(validation_reader) \
        .apply(tf.data.experimental.unbatch()) \
        .batch(batch_size_validation) \
        .map(lambda x: parser(x, validation_look_back, agg_relevant_columns_for_validation,
                                        batch_size_validation))
```

**Comment by selitvin** [comment]
I do not have a lot of experience with this kind of data distributions, so not sure if I can diagnose the bottleneck without playing with the code and the data (perhaps you can provide a code to construct a dataset with similar data distribution so I could repro?). 

These are some high level thoughts/questions:
- Excessive number of columns is indeed a suspect. I think I heard somewhere  that `libhdfs3` (unlike `libhdfs`) driver will issue a roundtrip request per column. Can you try passing `hdfs_driver="libhdfs"` to `make_batch_reader`.
- Curious if combining all columns into one column, just as an experiment will show speedup?
- Would try confirming that just a tight loop `for _ in make_batch_reader(...):` is approximately as slow as the whole setup with TF graph attached to it.
- How does your table look like? I see that you have almost 10K columns. How much rows in a rowgroup? What are the types of the fields? What is their size (in bytes?)

**Comment by jerrygb** [comment]
Thank you for the update @selitvin. Really appreciate your thoughts here.

Some information here,

- Seems that change the hdfs_driver, still did not quite improve the performance in this code path we discussed.
- So I do have a similar tf.data Dataset transformations, I am trying with csv data (ordered) and it seems have a slow performance as well. So it may not be related to petastorm after all. Additionally, since the data is ordered we did not have to pivot the columns and hence we are looking at only 400-500 feature columns.
- I did a rudimentary benchmark of just using the iterator, there by not attaching to the TF graph at all. It was taking around 500ms-1s per batch of sample data to produce the reshaped tensors.
- We were having 2 row groups. Feature and Label columns are only float32 tensors.

I was not doing any prefetch of the data. In spite of that, I still see a delay.

If you are aware of any tweaks or gotchas when using some of the tf.data module apis, like the ones mentioned on the code above. apply, batch, map, repeat, shuffle and so on.

**Comment by selitvin** [comment]
I think that iterating over `make_batch_reader` can not be faster than going through a TF graph that is connected to the same reader, so I would think that we need to focus the investigation around Petastorm code. 
From the petastorm side, we can play with the number of workers and the `reader_pool_type="thread"` or `reader_pool_type="process"`. Would suggest `pyarrow_serialize=True` with process worker pool implementation.

Can you run your benchmark under profiler? Maybe it would be clear what is the bottleneck, hence we could come up with ideas how to fix that?

Other than that, I would need to try to recreate a dataset with a similar data distribution to be able to toy with the code and see what behave differently from the expectation.

**Comment by austinzh** [comment]
The issue is in the unbatch and batch,  you can remove it and try. Just run a dummy iterator, you will see the huge different.
